### PR TITLE
Nh/android support ssl

### PR DESCRIPTION
**Pull request by nhachicha**
Fixing #4371 

This adds two options:

- `trustedRootCA` to provide the set of trusted root CA that Sync client should use

- `withoutSSLVerification` to disable OpenSSL certificate validation (for debug purpose or when using a self signed certificate in dev environment) 

TODO
- [x] changelog
- [x] tests

### CODE REVIEW

**Comment by zaki50** [comment]
This seems new feature.
Shouldn't this target `master`?

**Comment by zaki50** [comment]
Changed the target to `master`

File: realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp [hunk 1]
```diff
@@ -130,8 +132,19 @@ class JniConfigWrapper {
             user = SyncManager::shared().get_user(user_identity, refresh_token,
                                                   realm::util::Optional<std::string>(realm_auth_url));
         }
-        m_config.sync_config = std::make_shared<SyncConfig>(SyncConfig{
-            user, realm_url, SyncSessionStopPolicy::Immediately, std::move(bind_handler), std::move(error_handler)});
+        if (sync_ssl_trust_certificate_path) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp]: Any reason for this `if`. Couldn't the last parameter just be made optional?

File: realm/realm-library/src/main/java/io/realm/RealmCache.java [hunk 2]
```diff
@@ -15,6 +15,8 @@
  */
 package io.realm;
 
+import android.text.TextUtils;
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmCache.java]: Android import, please don't use this

File: realm/realm-library/src/main/java/io/realm/RealmCache.java [hunk 3]
```diff
@@ -527,55 +529,71 @@ synchronized void invokeWithLock(Callback0 callback) {
...
 
-                outputStream = new FileOutputStream(realmFile);
-                byte[] buf = new byte[4096];
-                int bytesRead;
-                while ((bytesRead = inputStream.read(buf)) > -1) {
-                    outputStream.write(buf, 0, bytesRead);
-                }
-            } catch (IOException e) {
+        // Copy Sync Server certificate path if available
+
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmCache.java]: Unnecessary whitespace?

File: realm/realm-library/src/main/java/io/realm/RealmCache.java [hunk 4]
```diff
@@ -527,55 +529,71 @@ synchronized void invokeWithLock(Callback0 callback) {
...
+        String syncServerCertificateAssetName = ObjectServerFacade.getFacade(configuration.isSyncConfiguration()).getSyncServerCertificateAssetName(configuration);
+        if (!TextUtils.isEmpty(syncServerCertificateAssetName)) {
+            String syncServerCertificateFilePath = ObjectServerFacade.getFacade(configuration.isSyncConfiguration()).getSyncServerCertificateFilePath(configuration);
+
+            // using getRealmDirectory avoid file collision between same filename from different users (Realms)
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmCache.java]: This doesn't seem to match the code?

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java [hunk 5]
```diff
@@ -185,13 +184,12 @@ boolean hasAssetFile() {
     }
 
     /**
-     * Returns input stream object to the Realm asset file.
+     * Returns the path to the Realm asset file.
      *
-     * @return input stream to the asset file.
-     * @throws IOException if copying the file fails.
+     * @return absolute path to the asset file.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java]: Is this true? Historically it is just the path for files in assets, and I believe that is a zipped file, so talking about "absolute path" doesn't feel correct?

File: realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java [hunk 6]
```diff
@@ -211,11 +211,13 @@ public static SharedRealm getInstance(RealmConfiguration config) {
...
-        String syncUserIdentifier = syncUserConf[0];
-        String syncRealmUrl = syncUserConf[1];
-        String syncRealmAuthUrl = syncUserConf[2];
-        String syncRefreshToken = syncUserConf[3];
+        Object[] syncUserConf = ObjectServerFacade.getSyncFacadeIfPossible().getUserAndServerUrl(config);
+        String syncUserIdentifier = (String) syncUserConf[0];
+        String syncRealmUrl = (String) syncUserConf[1];
+        String syncRealmAuthUrl = (String) syncUserConf[2];
+        String syncRefreshToken = (String) syncUserConf[3];
+        Boolean syncClientValidateSsl = (Boolean) syncUserConf[4];
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java]: why `Boolean` and not `boolean` ?

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 7]
```diff
@@ -17,6 +17,7 @@
 package io.realm;
 
 import android.content.Context;
+import android.text.TextUtils;
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: Please don't use this

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 8]
```diff
@@ -219,6 +234,37 @@ public boolean shouldDeleteRealmOnLogout() {
         return deleteRealmOnLogout;
     }
 
+    /**
+     * Returns the name of the certificate stored under the {@code assets}, to be used to validate
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: remove 'the'.

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 9]
```diff
@@ -596,6 +646,42 @@ public Builder errorHandler(SyncSession.ErrorHandler errorHandler) {
...
+        }
+
+        /**
+         * This will disable TLS certificate verification for the remote Realm Object Server.
+         * It is not recommended to use this in production.
+         * <p>
+         * This might be useful in non-production environments where you use a self-signed certificate
+         * for testing for example.
+         */
+        public Builder withoutSSLVerification() {
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: I much prefer `disableSSLVerification`

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 10]
```diff
@@ -596,6 +646,42 @@ public Builder errorHandler(SyncSession.ErrorHandler errorHandler) {
...
+            this.serverCertificateAssetName = filename;
+            return this;
+        }
+
+        /**
+         * This will disable TLS certificate verification for the remote Realm Object Server.
+         * It is not recommended to use this in production.
+         * <p>
+         * This might be useful in non-production environments where you use a self-signed certificate
+         * for testing for example.
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: Since you start with `this might` , i would probably just remove `for example`.

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 11]
```diff
@@ -733,6 +819,16 @@ public SyncConfiguration build() {
                 throw new IllegalStateException("Could not create directory for saving the Realm: " + realmFileDirectory);
             }
 
+            if (!TextUtils.isEmpty(serverCertificateAssetName)) {
+                if (!syncClientValidateSsl) {
+                    throw new IllegalStateException("SSL Verification is disable, server certificate should not be provided");
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: It feels a bit weird to me to have this check as well. The code for this would look like:

```
if (BuildType.DEBUG) {
  builder.disableSSLVerification()
} else {
  builder.trustedCA("myserver.pem")
}
```
Without this check it would be 
```
builder.trustedCA("myserver.pem")
if (BuildType.DEBUG) {
  builder.disableSSLVerification()
}
```
Which feels nicer to me?

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java [hunk 12]
```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2016 Realm Inc.
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java]: 2017

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+    public Timeout globalTimeout = Timeout.seconds(10);
+
+    @Rule
+    public final TestSyncConfigurationFactory configurationFactory = new TestSyncConfigurationFactory();
+
+    @Test
+    public void trustedRootCA() throws InterruptedException {
+        String username = UUID.randomUUID().toString();
+        String password = "password";
+        SyncUser user = SyncUser.login(SyncCredentials.usernamePassword(username, password, true), Constants.AUTH_URL);
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java]: It feels slightly weird that this will use the Java defined CA's instead of our custom one, but that is probably outside the scope of this PR.

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+
+        // 2. Local state should now be completely reset. Open the Realm again with a new configuration which should
+        // download the uploaded changes.
+        user = SyncUser.login(SyncCredentials.usernamePassword(username, password), Constants.AUTH_URL);
+        SyncConfiguration config = new SyncConfiguration.Builder(user, Constants.USER_REALM_SECURE)
+                .schema(StringOnly.class)
+                .build();
+        realm = Realm.getInstance(config);
+        try {
+            assertTrue(realm.isEmpty());
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java]: Isn't this reporteed as not working somewhere else, like through the ErrorHandler? It feels problematic that Sync just silently fails.

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncedRealmTests.java [hunk 15]
```diff
@@ -1,184 +0,0 @@
...
-
-    @Rule
-    public RunInLooperThread looperThread = new RunInLooperThread();
-
-    @Rule
-    public final UiThreadTestRule uiThreadTestRule = new UiThreadTestRule();
-
-    @Test
-    @UiThreadTest
-    public void waitForInitialRemoteData_mainThreadThrows() {
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncedRealmTests.java]: All these tests are in the other class, right?

File: realm/realm-library/src/main/java/io/realm/RealmCache.java (hunk 2 above)
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/RealmCache.java]: we're already using `TextUtils` in the code base ex: https://github.com/realm/realm-java/blob/9cea9356930b4357403b46ef4ca9995598a078ce/realm/realm-library/src/main/java/io/realm/RealmConfiguration.java#L184

Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmCache.java]: That is also a mistake 😉

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java (hunk 14 above)
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java]: We should file a change request against Sync then. That doesn't sem very user friendly.

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,3 +1,9 @@
+## 3.3.0
+
+### Enhancements
+
+* [ObjectServer] Added two options to `SyncConfiguration` to provide a trusted root CA `trustedRootCA` and to disable SSL validation `disableSSLVerification`. fixes (#4371).
```
Comment (cmelchior) [inline CHANGELOG.md]: I would just remove `fixes`. We haven't done that elsewhere.

File: realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp (hunk 1 above)
Comment (cmelchior) [inline realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp]: I would probably go with the `util:none` approach in this case, especially since the list of parameters are so long

File: realm/realm-library/src/main/java/io/realm/RealmCache.java (hunk 2 above)
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmCache.java]: Still present?

File: realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java (hunk 6 above)
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java]: what about `boolean syncClientValidateSsl = (syncUserConf[4] == Boolean.TRUE)`?

**Comment by nhachicha** [comment]
@realm/java

File: realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp [hunk 17]
```diff
@@ -130,8 +132,17 @@ class JniConfigWrapper {
...
+
+        util::Optional<std::string> ssl_trust_certificate_path = util::none;
+        if (sync_ssl_trust_certificate_path) {
+            ssl_trust_certificate_path =
+                realm::util::Optional<std::string>(JStringAccessor(env, sync_ssl_trust_certificate_path));
+        }
         m_config.sync_config = std::make_shared<SyncConfig>(SyncConfig{
-            user, realm_url, SyncSessionStopPolicy::Immediately, std::move(bind_handler), std::move(error_handler)});
+            user, realm_url, SyncSessionStopPolicy::Immediately, std::move(bind_handler), std::move(error_handler),
+            nullptr, util::none, sync_client_validate_ssl, ssl_trust_certificate_path});
```
Comment (beeender) [inline realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp]: is the `realm_encryption_key` newly added param to build `SyncConfig`? Don't we need to pass the key here?

File: realm/realm-library/src/main/java/io/realm/RealmCache.java [hunk 18]
```diff
@@ -527,55 +528,69 @@ synchronized void invokeWithLock(Callback0 callback) {
...
+        String syncServerCertificateAssetName = ObjectServerFacade.getFacade(configuration.isSyncConfiguration()).getSyncServerCertificateAssetName(configuration);
+        if (!Util.isEmptyString(syncServerCertificateAssetName)) {
+            String syncServerCertificateFilePath = ObjectServerFacade.getFacade(configuration.isSyncConfiguration()).getSyncServerCertificateFilePath(configuration);
+
+            File certificateFile = new File(syncServerCertificateFilePath);
+            copyFileIfNeeded(syncServerCertificateAssetName, certificateFile);
+        }
+    }
+
+    private static void copyFileIfNeeded(String assetFileName, File certificateFile) {
```
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/RealmCache.java]: name for `certificateFile` is confusing, this method is used for copying asset realm file as well.

File: realm/realm-library/src/main/java/io/realm/RealmCache.java [hunk 19]
```diff
@@ -527,55 +528,69 @@ synchronized void invokeWithLock(Callback0 callback) {
...
-                    outputStream.write(buf, 0, bytesRead);
-                }
-            } catch (IOException e) {
+        // Copy Sync Server certificate path if available
+        String syncServerCertificateAssetName = ObjectServerFacade.getFacade(configuration.isSyncConfiguration()).getSyncServerCertificateAssetName(configuration);
+        if (!Util.isEmptyString(syncServerCertificateAssetName)) {
+            String syncServerCertificateFilePath = ObjectServerFacade.getFacade(configuration.isSyncConfiguration()).getSyncServerCertificateFilePath(configuration);
+
+            File certificateFile = new File(syncServerCertificateFilePath);
+            copyFileIfNeeded(syncServerCertificateAssetName, certificateFile);
```
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/RealmCache.java]: corner case: what if user wants to change the cert file to a different one with the same name?

File: realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java [hunk 20]
```diff
@@ -232,7 +234,9 @@ public static SharedRealm getInstance(RealmConfiguration config, SchemaVersionLi
                 syncRealmUrl,
                 syncRealmAuthUrl,
                 syncUserIdentifier,
-                syncRefreshToken);
+                syncRefreshToken,
+                syncClientValidateSsl,
+                syncSslTrustCertificatePath);
```
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java]: 14 params for creating config :P 🎉 we may want to find a way to make this look better, but it is not the scope of this PR.

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 21]
```diff
@@ -596,6 +647,42 @@ public Builder errorHandler(SyncSession.ErrorHandler errorHandler) {
         }
 
         /**
+         * Provide the trusted root certificate(s) authority (CA) in {@code PEM} format, that should be used to
```
Comment (beeender) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: `Provide` -> `Provides`

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 22]
```diff
@@ -596,6 +647,42 @@ public Builder errorHandler(SyncSession.ErrorHandler errorHandler) {
...
+         * validate the TLS connections to the Realm Object Server.
+         * <p>
+         * The file should be stored under {@code assets}, it will be copied at runtime into the internal storage.
+         * <p>
+         * Note: This is similar to passing the parameter {@code CAfile} to {@code SSL_CTX_load_verify_locations},
+         *       Therefor it is recommended to include only the root CA you trust, and not the entire list of root CA
```
Comment (beeender) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: `Therefore`

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+        // make sure the changes gets to the server
+        SystemClock.sleep(TimeUnit.SECONDS.toMillis(2));  // FIXME: Replace with Sync Progress Notifications once available.
+        realm.close();
+        user.logout();
+        Realm.deleteRealm(configOld);
+
+        // 2. Local state should now be completely reset. Open the Realm again with a new configuration which should
+        // download the uploaded changes.
+        user = SyncUser.login(SyncCredentials.usernamePassword(username, password), Constants.AUTH_URL);
+        SyncConfiguration config = new SyncConfiguration.Builder(user, Constants.USER_REALM_SECURE)
```
Comment (beeender) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java]: why not using `configurationFactory.createSyncConfigurationBuilder`  here?

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
```
Comment (beeender) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java]: ditto

File: realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp (hunk 17 above)
Comment (nhachicha) [inline realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp]: yes setting the `realm_encryption_key` here should fix https://github.com/realm/realm-java/issues/4561 as well

File: realm/realm-library/src/main/java/io/realm/RealmCache.java (hunk 19 above)
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/RealmCache.java]: This is a similar use case with the existing [implementation](https://github.com/realm/realm-java/blob/deb25a87d08d93ebc77c5c8d9e81b8a632485b83/realm/realm-library/src/main/java/io/realm/RealmCache.java#L533) i.e what happen if you want to change the Realm shipped in the asset file with the same name. 

For now both use cases will need, the user to uninstall or clear internal data

File: realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java (hunk 20 above)
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java]: Kotlin to the help with named params! 😅
